### PR TITLE
UXD Fork Solend interactions + execute all proposals

### PR DESCRIPTION
**Pull request by ThomasProust**

### CODE REVIEW

File: .github/workflows/pull-request.yml [hunk 1]
```diff
@@ -1,16 +0,0 @@
...
-jobs:
-  test:
-    name: yarn test-all
-    runs-on: ubuntu-latest
-    steps:
-      - uses: actions/checkout@v2
-      - uses: actions/setup-node@v2
-        with:
-          node-version: lts/*
-          cache: yarn
```
Comment (0xCryptoSheik) [inline .github/workflows/pull-request.yml:14]: revert changes to this file

File: .gitignore [hunk 2]
```diff
@@ -33,5 +33,10 @@ yarn-error.log*
 # vercel
 .vercel
 
+# visual studio
+.vscode
+
```
Comment (0xCryptoSheik) [inline .gitignore:38]: revert changes to this file

File: .vscode/settings.json [hunk 3]
```diff
@@ -1,4 +0,0 @@
-{
-  "typescript.tsdk": "node_modules/typescript/lib",
```
Comment (0xCryptoSheik) [inline .vscode/settings.json:2]: revert changes to this file

File: components/instructions/ExecuteAllInstructionButton.tsx [hunk 4]
```diff
@@ -0,0 +1,147 @@
+import { useEffect, useState } from 'react'
+import {
+  InstructionExecutionStatus,
+  Proposal,
+  ProposalTransaction,
+  ProposalState,
+} from '@solana/spl-governance'
+import React from 'react'
+import { CheckCircleIcon, PlayIcon, RefreshIcon } from '@heroicons/react/solid'
```
Comment (0xCryptoSheik) [inline components/instructions/ExecuteAllInstructionButton.tsx:9]: group dependencies imports at the top, local imports below for better visibility

File: components/instructions/ExecuteAllInstructionButton.tsx [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+  const [currentSlot, setCurrentSlot] = useState(0)
+
+  const canExecuteAt = proposal?.account.votingCompletedAt
+    ? proposal.account.votingCompletedAt.toNumber() + 1
+    : 0
+
+  const ineligibleToSee = currentSlot - canExecuteAt >= 0
+
+  const rpcContext = new RpcContext(
+    new PublicKey(proposal.owner.toString()),
```
Comment (0xCryptoSheik) [inline components/instructions/ExecuteAllInstructionButton.tsx:53]: `proposal.owner` is already a `PublicKey`, right? Why need to re-create it ?

File: components/instructions/programs/raydium.tsx [hunk 6]
```diff
@@ -1,111 +0,0 @@
-import { Connection } from '@solana/web3.js'
-import { nu64, struct, u8 } from 'buffer-layout'
-import { AccountMetaData } from '@solana/spl-governance'
-import { fmtMintAmount } from '../../../tools/sdk/units'
-import { tryGetTokenMint } from '../../../utils/tokens'
-
-const RAYDIUM_AMM_INSTRUCTIONS = {
```
Comment (0xCryptoSheik) [inline components/instructions/programs/raydium.tsx:7]: Why delete this raydium instructions file as part of adding Solend?

File: components/instructions/programs/solend.tsx [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+      getDataUI: (
+        _connection: Connection,
+        _data: Uint8Array,
+        accounts: AccountMetaData[]
+      ) => {
+        // All accounts starting at index 2 are reserve accounts
+        const reserveAccounts = accounts.slice(2)
+
+        const reserveNames = reserveAccounts.map(
+          (x) =>
```
Comment (0xCryptoSheik) [inline components/instructions/programs/solend.tsx:54]: `x` -> `account` ?

File: components/instructions/programs/solend.tsx [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+            SolendConfiguration.getTokenNameByReservePublicKey(x.pubkey) ??
+            'unknown'
+        )
+
+        return (
+          <div style={{ display: 'flex', flexDirection: 'column' }}>
+            {reserveNames.map((x, xi) => (
+              <div
+                key={x}
```
Comment (0xCryptoSheik) [inline components/instructions/programs/solend.tsx:63]: `x` ?

File: components/instructions/programs/solend.tsx [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+                style={{ display: 'flex', justifyContent: 'space-between' }}
+              >
+                <span>Reserve #{xi + 1}</span>
+                <span>{x}</span>
```
Comment (0xCryptoSheik) [inline components/instructions/programs/solend.tsx:67]: `x` ?

File: components/instructions/programs/solend.tsx [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+        )
+
+        return (
+          <div style={{ display: 'flex', flexDirection: 'column' }}>
+            {reserveNames.map((x, xi) => (
+              <div
+                key={x}
+                style={{ display: 'flex', justifyContent: 'space-between' }}
+              >
+                <span>Reserve #{xi + 1}</span>
```
Comment (0xCryptoSheik) [inline components/instructions/programs/solend.tsx:66]: `xi` ?

File: components/instructions/programs/system.tsx [hunk 11]
```diff
@@ -1,10 +1,42 @@
-import { BN } from '@project-serum/anchor'
 import { Connection, PublicKey } from '@solana/web3.js'
+import { AccountMetaData } from '@solana/spl-governance'
+import { BN } from '@project-serum/anchor'
 import { getMintDecimalAmountFromNatural } from '@tools/sdk/units'
 import { tryGetMint } from '@utils/tokens'
 import { WSOL_MINT } from '../tools'
 import BufferLayout from 'buffer-layout'
 
+const SYSTEM_PROGRAM_ID = new PublicKey('11111111111111111111111111111111')
```
Comment (0xCryptoSheik) [inline components/instructions/programs/system.tsx:9]: Isn't there an existing export available in the app or dependencies for this?

File: hooks/useGovernanceAssets.ts [hunk 12]
```diff
@@ -142,7 +145,69 @@ export default function useGovernanceAssets() {
...
   )
+
+  const getGovernedMultiTypeAccounts = async (): Promise<
+    GovernedMultiTypeAccount[]
+  > => {
+    const mintWithGovernances = await getMintWithGovernances()
+
+    return governancesArray.map((gov) => {
+      const governedTokenAccount = governedTokenAccounts.find(
+        (x) => x.governance?.pubkey.toBase58() === gov.pubkey.toBase58()
```
Comment (0xCryptoSheik) [inline hooks/useGovernanceAssets.ts:156]: `PublicKey` has a `.equals(publicKey)` method that seems preferable

File: hooks/useGovernanceAssets.ts [hunk 13]
```diff
@@ -142,7 +145,69 @@ export default function useGovernanceAssets() {
...
+      )
+      if (governedTokenAccount) {
+        return governedTokenAccount as GovernedMultiTypeAccount
+      }
+
+      const mintGovernance = mintWithGovernances.find(
+        (x) => x.governance?.pubkey.toBase58() === gov.pubkey.toBase58()
```
Comment (0xCryptoSheik) [inline hooks/useGovernanceAssets.ts:163]: also still `x` everywhere

File: next.config.js [hunk 14]
```diff
@@ -9,6 +9,13 @@ const withTM = require('next-transpile-modules')([
 
 module.exports = withTM({
   webpack: (config, { isServer }) => {
+    config.resolve = {
+      ...config.resolve,
+      alias: {
+        ...config.resolve.alias,
+        '@project-serum/anchor$': '@project-serum/anchor/dist/esm',
```
Comment (0xCryptoSheik) [inline next.config.js:16]: This was only necessary to hotfix broken version of `@project-serum/anchor`, it can be removed now & `@project-serum/anchor` put back to a stable, recent version

File: next.config.js [hunk 15]
```diff
@@ -22,4 +29,13 @@ module.exports = withTM({
     DEVNET_RPC: process.env.DEVNET_RPC,
     DEFAULT_GOVERNANCE_PROGRAM_ID: process.env.DEFAULT_GOVERNANCE_PROGRAM_ID,
   },
+  async redirects() {
```
Comment (0xCryptoSheik) [inline next.config.js:32]: This is not needed here

File: package.json [hunk 16]
```diff
@@ -4,7 +4,7 @@
   "license": "MIT",
   "version": "1.0.0",
   "scripts": {
-    "dev": "next dev",
+    "dev": "next dev -p 3000",
```
Comment (0xCryptoSheik) [inline package.json:7]: This should be changed in this PR

File: package.json [hunk 17]
```diff
@@ -32,7 +33,7 @@
     "@marinade.finance/marinade-ts-sdk": "^2.0.9",
     "@metaplex/js": "^4.10.1",
     "@nfteyez/sol-rayz": "^0.8.0",
-    "@project-serum/anchor": "^0.10.0",
+    "@project-serum/anchor": "^0.19.0",
```
Comment (0xCryptoSheik) [inline package.json:36]: Should be updated to a stable version that doesn't require our workaround

File: package.json [hunk 18]
```diff
@@ -73,6 +75,7 @@
     "@types/jest": "^27.0.3",
     "@types/node": "^14.14.25",
     "@types/react": "^17.0.1",
+    "@types/react-virtualized": "^9.21.15",
```
Comment (0xCryptoSheik) [inline package.json:78]: This shouldn't be added here preferably

File: package.json [hunk 19]
```diff
@@ -84,8 +87,10 @@
     "jest": "^27.4.5",
     "lint-staged": "^10.0.10",
     "next-router-mock": "^0.6.3",
+    "patch-package": "^6.4.7",
```
Comment (0xCryptoSheik) [inline package.json:90]: no longer needed when we remove our workaround

File: package.json [hunk 20]
```diff
@@ -84,8 +87,10 @@
...
     "postcss": "^8.2.12",
     "postcss-preset-env": "^6.7.0",
+    "postinstall-postinstall": "^2.1.0",
```
Comment (0xCryptoSheik) [inline package.json:93]: no longer needed when we remove our workaround

File: pages/dao/[symbol]/proposal/components/instructions/CreateAssociatedTokenAccount.tsx [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+          handleSetForm({ value, propertyName: 'splTokenMintUIName' })
+        }
+        error={formErrors['baseTokenName']}
+      >
+        {Object.entries(SPL_TOKENS).map(([key, { name, mint }]) => (
+          <Select.Option key={key} value={name}>
+            <div style={{ display: 'flex', flexDirection: 'column' }}>
+              <span>{name}</span>
+
+              <span style={{ color: 'grey', fontSize: '0.8em' }}>
```
Comment (0xCryptoSheik) [inline pages/dao/[symbol]/proposal/components/instructions/CreateAssociatedTokenAccount.tsx:163]: project uses tailwind if I'm not mistaken, we should follow the existing convention

File: pages/dao/[symbol]/proposal/new.tsx [hunk 22]
```diff
@@ -55,6 +63,23 @@ export const NewProposalContext = createContext<InstructionsContext>(
...
+// Takes the first encountered governance account
+function extractGovernanceAccountFromInstructionsData(
+  instructionsData: ComponentInstructionData[]
+): ProgramAccount<Governance> | null {
+  return instructionsData.reduce((tmp, x) => {
+    if (tmp) {
+      return tmp
+    }
+
+    if (x.governedAccount) {
```
Comment (0xCryptoSheik) [inline pages/dao/[symbol]/proposal/new.tsx:75]: `return tmp ?? instructionData.governedAccount` ?

File: patches/@project-serum+anchor+0.19.0.patch [hunk 23]
```diff
@@ -0,0 +1,22 @@
+diff --git a/node_modules/@project-serum/anchor/dist/esm/program/common.js b/node_modules/@project-serum/anchor/dist/esm/program/common.js
+index 164d8ff..bab8614 100644
+--- a/node_modules/@project-serum/anchor/dist/esm/program/common.js
++++ b/node_modules/@project-serum/anchor/dist/esm/program/common.js
+@@ -38,15 +38,6 @@ export function validateAccounts(ixAccounts, accounts = {}) {
+ }
+ // Translates an address to a Pubkey.
+ export function translateAddress(address) {
```
Comment (0xCryptoSheik) [inline patches/@project-serum+anchor+0.19.0.patch:8]: Can be removed after update to a stable version of anchor

File: tools/routing.ts [hunk 24]
```diff
@@ -1,7 +1,10 @@
 import { RealmInfo } from 'models/registry/api'
 
 export function getRealmExplorerHost(realmInfo: RealmInfo | undefined) {
-  return realmInfo?.symbol === 'MNGO'
-    ? 'dao.mango.markets'
-    : 'realms-explorer.com'
+  switch (realmInfo?.symbol) {
+    case 'MNGO':
```
Comment (0xCryptoSheik) [inline tools/routing.ts:5]: Unnecessary change?

File: tools/sdk/solend/configuration.ts [hunk 25]
```diff
@@ -0,0 +1,137 @@
...
+  ): string | undefined {
+    return Object.entries(this.supportedMintsInformation).reduce(
+      (tmp, [mintName, { reserve }]) => {
+        if (reserveToFind.toString() === reserve.toString()) {
+          return mintName
+        }
+
+        return tmp
+      },
+      void 0
```
Comment (0xCryptoSheik) [inline tools/sdk/solend/configuration.ts:132]: I recognize this, please `undefined`, no human writes `void`

File: actions/executeInstructions.ts [hunk 26]
```diff
@@ -0,0 +1,54 @@
...
+    connection,
+    signers: [],
+  })
+
+  await sendSignedTransaction({
+    signedTransaction,
+    connection,
+    sendingMessage: 'Executing instruction',
+    successMessage: 'Execution finalized',
+    timeout: DEFAULT_TIMEOUT,
```
Comment (mschneider) [inline actions/executeInstructions.ts:52]: DEFAULT_TIMEOUT  constant can be removed, usually better if all timeouts are controlled centrally

File: components/instructions/ExecuteAllInstructionButton.tsx [hunk 27]
```diff
@@ -0,0 +1,147 @@
...
+  useEffect(() => {
+    if (ineligibleToSee && proposal) {
+      const timer = setTimeout(() => {
+        rpcContext.connection.getSlot().then(setCurrentSlot)
+      }, 5000)
+
+      return () => {
+        clearTimeout(timer)
+      }
+    }
```
Comment (mschneider) [inline components/instructions/ExecuteAllInstructionButton.tsx:69]: not sure what this is about, maybe add a comment why it is needed

File: components/instructions/ExecuteAllInstructionButton.tsx [hunk 28]
```diff
@@ -0,0 +1,147 @@
...
+  }, [ineligibleToSee, rpcContext.connection, currentSlot])
+
+  const onExecuteInstructions = async () => {
+    setPlaying(PlayState.Playing)
+
+    try {
+      await executeInstructions(rpcContext, proposal, proposalInstructions)
+      await fetchRealm(realmInfo?.programId, realmInfo?.realmId)
```
Comment (mschneider) [inline components/instructions/ExecuteAllInstructionButton.tsx:77]: not sure you need to await fetchRealm
